File: Upupu/CameraController.m
//
//  CameraController.m
//  Upupu
//
//  Created by Takashi AOKI on 4/22/11.
//  Copyright 2011 Xcoo, Inc. All rights reserved.
//

#import "CameraController.h"

#import "CameraViewController.h"
#import "UploadViewController.h"
#import "IASKAppSettingsViewController.h"

#import "DropboxUploader.h"

@interface CameraController() <CameraViewControllerDelegate, UploadViewControllerDelegate, IASKSettingsDelegate>

@property (nonatomic, retain) CameraViewController *cameraViewController;
@property (nonatomic, retain) UploadViewController *uploadViewController;

@end

@implementation CameraController

- (id)init
{
    self = [super init];
    
    if ( !self ) {
        return nil;
    }
    
    self.cameraViewController = [[CameraViewController alloc] initWithNibName:@"CameraViewController"
                                                                       bundle:nil];
    _cameraViewController.delegate = self;
    
    self.uploadViewController = [[UploadViewController alloc] initWithNibName:@"UploadViewController"
                                                                       bundle:nil];
    _uploadViewController.delegate = self;
        
    self.navigationBar.barStyle = UIBarStyleBlackOpaque;
    self.navigationBar.hidden = YES;
    
    [self pushViewController:_cameraViewController animated:NO];
    
    return self;
}

- (void)didReceiveMemoryWarning
{
    // Releases the view if it doesn't have a superview.
    [super didReceiveMemoryWarning];
    
    // Release any cached data, images, etc that aren't in use.
}

#pragma mark - View lifecycle

- (void)viewDidLoad
{
    [super viewDidLoad];
}

- (void)viewDidUnload
{
    SAFE_RELEASE(_cameraViewController)
    SAFE_RELEASE(_uploadViewController)
    
    [super viewDidUnload];
}

- (void) dealloc
{
    SAFE_RELEASE(_cameraViewController)
    SAFE_RELEASE(_uploadViewController)
    
    [super dealloc];
}

- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation
{
    return interfaceOrientation == UIInterfaceOrientationPortrait;
}

#pragma mark - CameraViewControllerDelegate -

-(void) cameraViewController:(UIViewController *)viewController didFinishedWithImage:(UIImage *)image
{
    _uploadViewController.image = image;
    
    CameraViewController *cameraViewController = (CameraViewController *)viewController;
    _uploadViewController.savePhotoAlbum = !cameraViewController.isSourcePhotoLibrary;
    
    [[UIApplication sharedApplication] setStatusBarHidden:NO withAnimation:UIStatusBarAnimationSlide];
    [UIApplication sharedApplication].statusBarStyle = UIStatusBarStyleLightContent;
    
    [self pushViewController:_uploadViewController animated:YES];
}

#pragma mark - UploadViewControllerDelegate -

- (void) uploadViewControllerDidReturn:(UIViewController *)controller
{
    [[UIApplication sharedApplication] setStatusBarHidden:YES withAnimation:UIStatusBarAnimationSlide];
    [UIApplication sharedApplication].statusBarStyle = UIStatusBarStyleDefault;
    
    [self popViewControllerAnimated:YES];
}

- (void) uploadViewControllerDidFinished:(UIViewController *)controller
{
    [[UIApplication sharedApplication] setStatusBarHidden:YES withAnimation:UIStatusBarAnimationSlide];
    [UIApplication sharedApplication].statusBarStyle = UIStatusBarStyleDefault;
    
    [self popViewControllerAnimated:YES];
}

- (void) uploadViewControllerDidSetup:(UIViewController *)controller
{
    IASKAppSettingsViewController *settingsController = [[IASKAppSettingsViewController alloc] init];
    settingsController.delegate = self;
    settingsController.showCreditsFooter = NO;
    
    UINavigationController *navController = [[UINavigationController alloc] initWithRootViewController:settingsController];
    navController.modalTransitionStyle = UIModalTransitionStyleFlipHorizontal;
    [self presentViewController:navController animated:YES completion:^{
        [[UIApplication sharedApplication] setStatusBarHidden:NO withAnimation:UIStatusBarAnimationFade];
        [[UIApplication sharedApplication] setStatusBarStyle:UIStatusBarStyleDefault animated:YES];
    }];
    
    [navController release];
    [settingsController release];
}

#pragma mark - IASKSettingsDelegate -

- (void)settingsViewControllerDidEnd:(IASKAppSettingsViewController *)sender 
{
    [[UIApplication sharedApplication] setStatusBarHidden:NO withAnimation:UIStatusBarAnimationFade];
    [[UIApplication sharedApplication] setStatusBarStyle:UIStatusBarStyleLightContent animated:YES];
    [sender dismissViewControllerAnimated:YES completion:nil];
}

- (void)settingsViewController:(IASKAppSettingsViewController *)sender buttonTappedForKey:(NSString *)key 
{
	if ([key isEqualToString:@"dropbox_link_pref"]) {
        [[DropboxUploader sharedInstance] linkFromController:sender];
    }
    
    [sender dismissViewControllerAnimated:YES completion:nil];
}

@end

File: Upupu/Controllers/CameraController.swift
//
//  CameraController.swift
//  Upupu
//
//  Created by Toshiki Takeuchi on 8/29/16.
//  Copyright © 2016 Xcoo, Inc. All rights reserved.
//  See LISENCE for Upupu's licensing information.
//

import UIKit

import InAppSettingsKit

class CameraController: UINavigationController, CameraViewControllerDelegate,
UploadViewControllerDelegate, IASKSettingsDelegate {

    private var cameraViewController: CameraViewController!
    private var uploadViewController: UploadViewController!

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        initialize()
    }

    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: NSBundle?) {
        super.init(nibName: nil, bundle: nil)
        initialize()
    }

    convenience init() {
        self.init(nibName: nil, bundle: nil)
    }

    private func initialize() {
        cameraViewController = CameraViewController(nibName: "CameraViewController", bundle: nil)
        cameraViewController.delegate = self

        uploadViewController = UploadViewController(nibName: "UploadViewController", bundle: nil)
        uploadViewController.delegate = self

        navigationBar.barStyle = UIBarStyle.BlackOpaque
        navigationBar.hidden = true

        pushViewController(cameraViewController, animated: false)
    }

    override func viewDidLoad() {
        super.viewDidLoad()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }

    override func shouldAutorotate() -> Bool {
        return UIDevice.currentDevice().orientation == .Portrait
    }

    // MARK: - CameraViewControllerDelegate

    func cameraViewController(cameraViewController: CameraViewController,
                              didFinishedWithImage image: UIImage?) {
        uploadViewController.image = image
        uploadViewController.shouldSavePhotoAlbum = !cameraViewController.isSourcePhotoLibrary

        let application = UIApplication.sharedApplication()
        application.setStatusBarHidden(false, withAnimation: .Slide)
        application.statusBarStyle = .LightContent

        pushViewController(uploadViewController, animated: true)
    }

    // MARK: - UploadViewControllerDelegate

    func uploadViewControllerDidReturn(uploadViewController: UploadViewController) {
        dispatch_async(dispatch_get_main_queue()) {
            let application = UIApplication.sharedApplication()
            application.setStatusBarHidden(true, withAnimation: .Slide)
            application.statusBarStyle = .Default
        }

        popViewControllerAnimated(true)
    }

    func uploadViewControllerDidFinished(uploadViewController: UploadViewController) {
        dispatch_async(dispatch_get_main_queue()) {
            let application = UIApplication.sharedApplication()
            application.setStatusBarHidden(true, withAnimation: .Slide)
            application.statusBarStyle = .Default
        }

        popViewControllerAnimated(true)
    }

    func uploadViewControllerDidSetup(uploadViewController: UploadViewController) {
        let settingsViewController = IASKAppSettingsViewController()
        settingsViewController.delegate = self
        settingsViewController.showCreditsFooter = false

        let navigationContoller = UINavigationController(rootViewController: settingsViewController)
        navigationContoller.modalTransitionStyle = .FlipHorizontal

        presentViewController(navigationContoller, animated: true) {
            let application = UIApplication.sharedApplication()
            application.setStatusBarHidden(false, withAnimation: .Fade)
            application.setStatusBarStyle(.Default, animated: true)
        }
    }

    // MARK: - IASKSettingsDelegate

    func settingsViewControllerDidEnd(sender: IASKAppSettingsViewController) {
        let application = UIApplication.sharedApplication()
        application.setStatusBarHidden(false, withAnimation: .Fade)
        application.setStatusBarStyle(.LightContent, animated: true)
        sender.dismissViewControllerAnimated(true, completion: nil)
    }

    func settingsViewController(sender: IASKAppSettingsViewController,
                                buttonTappedForSpecifier specifier: IASKSpecifier) {
        if specifier.key() == "dropbox_link_pref" {
            DropboxUploader.sharedInstance.linkFromController(sender)
        }

        sender.dismissViewControllerAnimated(true, completion: nil)
    }

}
